### PR TITLE
C#: Framework dependency detection.

### PR DESCRIPTION
**Pull request by michaelnebel**
In this PR we introduce assets.json unit tests for six representative target frameworks (that behaves slightly differently when generating assets files and how to list dependencies).
The frameworks are: `net70`, `net4.8`, `netstandard2.1`, `netstandard1.6`, `netcoreapp2.0`, `netcoreapp3.1`

As an outcome of introducing the tests, we also introduce the following improvements/modifications to the assets parser.
- Inspecting the `frameworkReferences` section of the `project.assets.json` files to identify which frameworks are being used (this is relevant for some of the target frameworks from the above list - but not all).
- If a *package* from the `target` section in the `project.assets.json` file contains a `compile` section we only include what is in the compile section (also for frameworks). If a framework doesn't have a `compile` section we do still fallback to including the entire framework package as a dependency.
- References to the empty `_._` placeholder files are ignored.

Furthermore,
- Framework package names have been re-factored into a separate class (to avoid duplication).

The changes in this PR shows some minor improvements on the dependency detection for the following projects (we are narrowing even further what the standalone extractor detects as a dependency).
Before
```
Results for tobyash86/WebGoat.NET
 Query Dependencies.ql:
  Result count - both/traced/standalone: 367/0/2

Results for AAEmu/AAEmu
 Query Dependencies.ql:
  Result count - both/traced/standalone: 237/0/4
```

After
```
Results for tobyash86/WebGoat.NET
 Query Dependencies.ql:
  Result count - both/traced/standalone: 367/0/0
Query Dependencies.ql:
  Result count - both/traced/standalone: 237/0/1
```

As a follow up PR, we should try to remove the logic in the dependency manager that always attempts to add a framework package. We still need to retain the logic that removes references to frameworks based on the framework priority order as we could risk different project files having different target frameworks (which might lead to a conflict in the information produced in the asset files).

### CODE REVIEW

**Comment by michaelnebel** [comment]
DCA shows
- No change in alerts
- No change in performance
- Only minor changes to the number of extractor errors and compilation errors.

As such this is not holistically detected as an improvement (or regression) by DCA, but I think that overall this makes the dependency logic more accurate and "easier" to incrementally improve further.

**Comment by tamasvajk** [comment]
Have you looked into the `cs/compilation-error` changes in DCA? What is the reason for the increase in count? It seems like there's no repo where the count decreased, but there are 5 repos, where it increased.

**Comment by michaelnebel** [comment]
> Have you looked into the `cs/compilation-error` changes in DCA? What is the reason for the increase in count? It seems like there's no repo where the count decreased, but there are 5 repos, where it increased.

I have investigated this for `NLog/NLog`.
On main we have:
```
Query Dependencies.ql:
  Result count - both/traced/standalone: 191/19/57
``
On this branch we have
```
 Query Dependencies.ql:
  Result count - both/traced/standalone: 191/19/53
```
By comparing the dependencies, the logic in the feature branch now doesn't include the following dll's
```
Microsoft.Identity.Client
System.Runtime.Caching
System.Configuration.ConfigurationManager
System.Security.Cryptography.ProtectedData
```
There are around 30 new compilation errors, which is mostly concerned about missing attributes (almost all "new" compilation errors are related to a single file). One example is:
<img width="483" alt="image" src="https://github.com/github/codeql/assets/6472811/20476428-64e7-444d-994a-bdb68cf58dd2">

which is located in `System.Web.Services`. Both on main and on this feature branch - there is already a compilation error saying that the `Services` namespace can't be found in `System.Web`. I just think that for some reason Roslyn provides more detailed errors than on main (in both cases it has already discoved that it can't find the Services).

**Comment by michaelnebel** [comment]
The project `DeafLight/SilentLux.AutoFixture.Moq.Xunit2` is completely broken on both main and the feature branch.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FrameworkPackageNames.cs [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+
+        public static string AspNetCoreFramework =>
+            "microsoft.aspnetcore.app.ref";
+
+        public static string WindowsDesktopFramework =>
+            "microsoft.windowsdesktop.app.ref";
+
+        public static readonly IEnumerable<string> AllFrameworks =
+            NetFrameworks
+                .Union(new string[] { AspNetCoreFramework, WindowsDesktopFramework });
```
Comment (tamasvajk) [inline csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FrameworkPackageNames.cs:25]: Minor: there's a mix of fields and properties here. Is there any reason for this?

**Review by tamasvajk — DISMISSED** [review]
Looks good to me.

File: csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FrameworkPackageNames.cs (hunk 1 above)
Comment (michaelnebel) [reply csharp/extractor/Semmle.Extraction.CSharp.DependencyFetching/FrameworkPackageNames.cs:25]: No, I will change that to use property initialisers.

**Comment by michaelnebel** [comment]
@tamasvajk : The only change since your last review is the last commit.

**Review by tamasvajk — APPROVED** [review]
LGTM.

**Pull request merged**